### PR TITLE
Use bulk merging of TDigestHistogram

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/OperatorStats.java [hunk 1]
```diff
@@ -538,7 +540,7 @@ public OperatorStats add(Iterable<OperatorStats> operators)
             OperatorInfo info = operator.getInfo();
             if (base != null && info != null) {
                 verify(base.getClass() == info.getClass(), "Cannot merge operator infos: %s and %s", base, info);
-                base = mergeInfo(base, info);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:541]: which `OperatorInfo` contain `TDigestHistogram`?

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/OperatorStats.java:541]: DirectExchangeClientStatus

File: core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java [hunk 2]
```diff
@@ -535,7 +534,7 @@ public StageInfo getStageInfo(Supplier<Iterable<TaskInfo>> taskInfosSupplier)
             inputBlockedTime += taskStats.getInputBlockedTime().roundTo(NANOSECONDS);
 
             bufferedDataSize += taskInfo.getOutputBuffers().getTotalBufferedBytes();
-            outputBufferUtilization = Optionals.combine(outputBufferUtilization, taskInfo.getOutputBuffers().getUtilization(), TDigestHistogram::mergeWith);
+            taskInfo.getOutputBuffers().getUtilization().ifPresent(bufferUtilizationHistograms::add);
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java:537]: Will this not increase peak memory usage, as we will have for a short period of time more TDigestHistograms in memory?

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientStatus.java [hunk 3]
```diff
@@ -162,6 +162,47 @@ public DirectExchangeClientStatus mergeWith(DirectExchangeClientStatus other)
...
+        long maxBufferedBytes = this.maxBufferedBytes;
+        long averageBytesPerRequest = this.averageBytesPerRequest;
+        long successfulRequestsCount = this.successfulRequestsCount;
+        int bufferedPages = this.bufferedPages;
+        int spilledPages = this.spilledPages;
+        long spilledBytes = this.spilledBytes;
+        boolean noMoreLocations = this.noMoreLocations;
+        ImmutableList.Builder<TDigestHistogram> requestDurations = ImmutableList.builderWithExpectedSize(others.size());
+        for (DirectExchangeClientStatus other : others) {
+            bufferedBytes = (bufferedBytes + other.bufferedBytes) / 2; // this is correct as long as all clients have the same buffer size (capacity)
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClientStatus.java:182]: If both values are the same, do we need to do something with them?
cannot just do:
```java
bufferedBytes = this.bufferedBytes;
```
outside loop?

File: core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java (hunk 2 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java:537]: yes, that is the trade-off of bulk processing vs two at a time processing, I don't think this requires a lot of memory or will stay in memory for long.

**Review by radek-starburst — APPROVED** [review]
lgtm

**Review by Dith3r — APPROVED** [review]

**Review by sopel39 — APPROVED** [review]

**Pull request merged**